### ISSUE TITLE
Ability to cancel script execution

### ISSUE THREAD

**Issue by m4ksio** [onflow/cadence#1435]
### Issue To Be Solved

It should be possible to cancel the execution of the script. Once usecase would be to time-out script execution on access nodes.

### Suggested Solution

Add standard Go context ( https://pkg.go.dev/context ) to ExecuteScript method in Runtime

**Comment by bluesign** [comment]
@m4ksio  Why would access node time-out my script execution ? Isn't gas limit better for that?

**Comment by m4ksio** [comment]
> @m4ksio Why would access node time-out my script execution ? Isn't gas limit better for that?

Gas-based limits are useful when we care about fairness. But sometimes all we care about is execution time. Also our gas model doesn't really work that great yet.

**Comment by turbolent** [comment]
@bluesign Script execution is a feature of the protocol, but concrete availability depends on the node operators. As script execution is computation, and computation costs money, operators may limit the amount through various means, one which could be time based.

**Comment by bluesign** [comment]
> @bluesign Script execution is a feature of the protocol, but concrete availability depends on the node operators. As script execution is computation, and computation costs money, operators may limit the amount through various means, one which could be time based.  

This can have some bad consequences like: 

- When I execute script, it can sometimes success and sometimes fail ( depending on the AN load balancer forwarded me )

- what if an AN decides to not run scripts at all or geo block some IPs

I think script execution is almost as important as transaction execution, some reliability there is very important

**Comment by turbolent** [comment]
cc @ramtinms

**Comment by ramtinms** [comment]
@bluesign I see your points and there are no double scripts are as important as transactions in flow ecosystem.

I think one important aspect is having a way to pass control signals from upper layers to cadence is really valuable and we could save resources when the output is going to be ignored, even if extra timeouts are not added by the AN or ENs, client GRPC calls always have a timeout or connections might always be dropped. A common case that has happened before is a grpc call gets a client timeout on a really short default timeout and the client constantly retries many times causing an unnecessary load on ENs (while the client would ignore the earlier calls anyway) or the attacker sends requests and drops the connection immediately.  

Even if we decide to add a script execution timeout to the Execution nodes (which I think is outside of the scope of this request), we are aiming to cover 99.999% of scripts and mostly keep time outs as a safety measure for attacks that figure out a way to keep their gas usage low but has a huge time and resource impact on ENs.

**Comment by bluesign** [comment]
Thanks @ramtinms, I totally understand the technical reason, and it is fair enough. 

> but concrete availability depends on the node operators.
to be honest my problem is this, on one hand I am hearing something as 'staked AN' terminology, then it is kinda conflicting in my brain with this statement. 

> we are aiming to cover 99.999% of scripts
maybe you will, but what is the motivation for node operator? 

But to be honest it will be so much clear, if you  ( as Flow Team ) can share a bit about what is in your head in the end game? Script execution, execution data etc, how this will fit in the big picture, in a byzantine proof way, as I am unable to see. 

Actually I am constantly asking some information about big picture ( end game ) for almost 1 year 

As a note: 

> attacks that figure out a way to keep their gas usage low but has a huge time and resource impact on ENs 

this is also not making sense for me, then something is wrong with gas calculation.  ( Though this discussion is another parallel I am really trying to avoid )

**Comment by j1010001** [comment]
To do this we need to add some mechanism to make the interpreter stoppable.

**Comment by j1010001** [comment]
Doable in the interpreter, might be easier to do if/when we have a VM.

**Comment by bluesign** [comment]
Doesn't AN already timeout script execution? Is this for some kind of clean stop?

**Comment by turbolent** [comment]
@bluesign Yes, it already does it, but indirectly through metering callbacks. This issue is about adding proper support for cancellation at any point.
